memory_w25q1128jv_spi/memory_w25q1128jv_spi.cc

```diff
@@ -0,0 +1,223 @@
+    if (address > 0xFFFF) return 0;
+
+    // Bundles Telemetry data struct into a 4 element array
+    // Telemetry data is written into the array until all 4 elements are full
+    if (Data_Bundle_Size < 3) {
+        Data_Bundle[Data_Bundle_Size] = data;
+        Data_Bundle_Size++;
+        return 1;
+    }
+
+    // resets bundled data variable and inserts the last Telemetry data values
+    Data_Bundle[3] = data;
+    Data_Bundle_Size = 0;
+
+    // Chip enable instruction command
+    uint8_t chip_enable[1] = {0x06};
+    // Page Program instruction command
+    uint8_t Page_program[4] = {0x02, (uint8_t)(address >> 8), (uint8_t)(address), 0x00};  // instruction opcode for Page Program
+
+    // chip enable instruction to enable chip
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(spi, chip_enable, 1, SPI_Timeout) != HAL_OK) return 0;
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_SET);
+
+    // Checks and waits until status register write bit  is = 0
+    // waits until Chip enable is done
+    if (MemoryW25q1128jvSpi::Read_Status_Reg1(2, 1) == 0) return 0;
+
+    // Write Data bundle into memory
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(spi, Page_program, 4, SPI_Timeout) != HAL_OK) return 0;
+    if (HAL_SPI_Transmit(spi, (uint8_t *)Data_Bundle, 256, SPI_Timeout) != HAL_OK) return 0;
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_SET);
+
+    // Checks and waits until status register busy bit  is = 0
+    // waits until page write is done
+    if (MemoryW25q1128jvSpi::Read_Status_Reg1(1, 0) == 0) return 0;
+
+    // increments to the next page address
+    address++;
+
+    return 1;
+}
+
+MemoryW25q1128jvSpi::afsTelemetryData *MemoryW25q1128jvSpi::Chip_Read(uint32_t read_address) {
+    // making sure that control pins are always high
+    HAL_GPIO_WritePin(wpPort, wpPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(holdPort, holdPin, GPIO_PIN_SET);
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_SET);
+
+    // initalisation of input, output, and dumby telemtrydata variable for temporary writing to DataBundle
+    uint8_t Read_Data[4] = {0x03, (uint8_t)(read_address >> 8), (uint8_t)(read_address), 0x00};
+    uint8_t out[256] = {0};
+    MemoryW25q1128jvSpi::afsTelemetryData Data;
+
+    // Reads address page in memory
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_RESET);
+    if (HAL_SPI_Transmit(spi, Read_Data, 4, SPI_Timeout) != HAL_OK) return Data_Bundle;
+    if (HAL_SPI_Receive(spi, out, 256, SPI_Timeout) != HAL_OK) return Data_Bundle;
+    HAL_GPIO_WritePin(csPort, csPin, GPIO_PIN_SET);
+
+    // Checks and waits until status register busy bit  is = 0
+    // waits until page read is done
+    if (MemoryW25q1128jvSpi::Read_Status_Reg1(0x01, 1) == 0) return 0;
+
+    // organisation of Databundle from read page until Data_Bundle array
+    for (int i = 0; i < 4; i++) {
+        // Specifier for telemetry data
+        Data.type = out[(64 * i)];
```

Comment:
Instead of doing these offset calculations, you should just cast `out` to a `afsTelemetryData` struct. Since that struct is packed, and assuming you have the fields in the right order, it will automatically fill everything correctly. The syntax would be something like this: 
```
Memory::afsTelemetryData* data = (Memory::afsTelemetryData*) out;
for(int i=0;i<4;i++) {
    dataBundle[i] = data[i];
}
```
If you really wanna be a C++ wizard and save memory, you could just read the data directly into Data_Bundle like this:
```
uint8_t* out = (uint8_t*)Data_Bundle;
HAL_SPI_Receive(spi, out, 256, blahblahblah

```